Induction principle example and doc (#2645)
dafny-lang/dafny: Test/tutorial/induction-principle.md

```

This theorem can be deduced from the following intermediate lemma, that we prove by
induction on the expression `e`:

```dafny
lemma InterpExpr_Pure(e: Expr, locals: set<string>, ctx: Context)
  requires IsPure(e, locals)
  requires ctx.Keys >= locals
  ensures
    match InterpExpr(e, ctx)
      case Success((_, ctx')) =>
        && ctx'.Keys == ctx.Keys
        // We only update variables captured by outer let-bindings
        && forall x | x in ctx.Keys && x !in locals :: ctx'[x] == ctx[x]
      case Failure => true
```

Let's consider the `Seq` case. In our AST, we define sequences as follows:

```dafny
datatype Expr =
  | Seq(e1: Expr, e2: Expr)
  | ... // Omitted
```

And the interpreter is defined on sequences as follows:

```dafny
function method InterpExpr(e: Expr, ctx: Context): Result<(int, Context)> {
  match e
    case Seq(e1, e2) =>
      var (_, ctx1) :- InterpExpr(e1, ctx); // Evaluate the first expression
      InterpExpr(e2, ctx1) // Continue

    // Omitted
    case ... =>
      ...
}
```

We face the problem that the proof of the `Seq` case requires introducing the same
intermediate values as the ones manipulated by the interpreter (like `ctx1`), so that we

> can *write* the recursive calls to `InterpExpr_Pure_WithLocals` (which give us our

Reviewer: What is this lemma? How does it relate to `InterpExpr` ? I'm not sure I understand the logic here. It uses "..." for the spec, but it was never defined before.

Author: Thanks for this! I renamed ands merged some predicates just before the PR, and missed the associated lemmas. I updated the document accordingly (here, the reference should be to `InterpExpr_Pure`, which is declared just above, hence the `...`).

be a big hassle, while the introduction of the induction principle drastically diminished
the pain.

<a id="pair-of-state"></a>
## A different kind of proof: manipulating a pair of states

So far, we only studied proofs which require using one context. This changes if we add
closures. One way of adding support for closures is to introduce a `Value` datatype,
where closures are defined by the context they capture, a sequence of input variables,
and their body:

```dafny
datatype Value
| Integer(i: int)
| Closure(ctx: Context, vars: seq<string>, body: Expr)
```

The problem when verifying compiler transformations is that we can't use Dafny's native
equality anymore, because local rewritings might modify the body of closures and hence their
values: we can't reason with a *syntactic* equality. We thus need to introduce a *semantic*
equality over values. This semantic equality would state that two closures are equal if,
when called on equal inputs (in the sense of this semantic equality), they evaluate
to equal results (in the sense of this semantic equality, again), in the spirit of the
following predicate:

```dafny

> predicate EqValue(cv: Value, cv': Closure)

Reviewer: ```suggestion
predicate EqValue(cv: Value, cv': Value)
  requires cv.Closure? && cv'.Closure?
```
?

Author: Thanks for noticing the typo! I renamed the predicate to `EqValue_Closure`, as its purpose is to illustrate the equivalence relation for the closures case (and consistently changed the types to `Closure`)

to be particularly useful when working on
[Dafny-in-Dafny](https://github.com/dafny-lang/compiler-bootstrap),
which we will use as a motivating example. The full code development can be found in the
[`induction-principle`](induction-principle/) folder. The development with the preserved git
history can be found in the Dafny-in-Dafny project, [here](https://github.com/dafny-lang/compiler-bootstrap/blob/639b29dab5b5f5e019d18145b513ab1b9b7fde7a/doc/proofs/induction-principle.md).

## Problem: proving lemmas about an interpreter

In the context of verifying a compiler, we need to introduce a semantics for the compiled
language. We can do so by defining a function `InterpExpr(e: Expr, st: State):
Result<(Value, State), Error>`, which given an expression and an initial state either
returns an error (division by zero, access to an unbound variable, etc.) or returns the
value this expression evaluates to, together with the new state in which the execution
terminates.

Verifying the compiler then requires us to repeatedly state and prove lemmas about this
`InterpExpr` function. For instance, given a transformation `TR: (Expr) -> Expr`, we might

> express the correctness of the transformation in the following way (which says that the new program fails no more often that

Reviewer 2: ```suggestion
express the correctness of the transformation in the following way (which says that the new program fails no more often than
```

the original one, and evaluates to the same result):

```dafny
forall e, st :: InterpExpr(e, st).Success? ==> InterpExpr(e, st) == InterpExpr(TR(e), st)
```

Such a lemma has to be proved over `e` or over the execution of the interpreter (that is, only some
lemmas can be proved by induction over `e`) for all the transformations we wish to
verify, and its proof generally follows very closely the structure of `InterpExpr`. This is
for instance the case for the following local rewriting:

```dafny
0 * e  -->  0       when e is pure
```

Moreover, verifying this rewriting requires us to prove the following auxiliary lemma (for the appropriate
definition of `IsPure`) which, again, mentions `InterpExpr`:

```dafny
forall e, st ::
  IsPure(e) ==>
    match InterpExpr(e, st)
      case Success((_, st')) => st' == st // state is unchanged
      case Failure => true
```

> **Note:** we mention above that not all lemmas can be proved by induction over the syntax.

Reviewer 2: ```suggestion
**Note:** We mention above that not all lemmas can be proved by induction over the syntax.
```

It is for instance the case if the AST contains loops, in which situation we often need to
manipulate a notion of fuel when defining the interpreter: the lemmas then require doing
the induction also on the fuel. Let's assume we can always do the induction on the syntax only

> (and in particular that don't have loops).

Reviewer 2: ```suggestion
(and in particular that there are no loops).
```

The issue we encounter here is that we have a lot of lemmas to prove about a single
function, `InterpExpr`. However, whenever we need to prove such a lemma in Dafny, we need
more or less to copy/paste the whole `InterpExpr` definition because the structure of the
proof follows the structure of this function. This process is time consuming,
leads to big proofs which are hard to maintain, and is error prone as we are blind while
instantiating function calls and lemmas calls (in particular, it is very easy to mix up
variable indices).

For instance, let's look at the correctness proof for the ``IsPure`` predicate, that we introduce
below:

```dafny
predicate method IsPure(e: Expr, locals: set<string> := {})
  decreases e
  // `locals`: the set of local variables which have been introduced by outer
  // let-bindings (if the expression only updates those variables then it is
  // pure, because those variables don't escape the scope of the term).
{
  match e
    case Var(name) => true
    case Literal(n) => true
    case Bind(bvar, bval, body) =>

>       IsPure(bval, locals)

Reviewer 2: ```suggestion
      && IsPure(bval, locals)
```

  match e
    case Seq(e1, e2) =>
      var (_, ctx1) :- InterpExpr(e1, ctx); // Evaluate the first expression
      InterpExpr(e2, ctx1) // Continue

    // Omitted
    case ... =>
      ...
}
```

We face the problem that the proof of the `Seq` case requires introducing the same
intermediate values as the ones manipulated by the interpreter (like `ctx1`), so that we
can *write* the recursive calls to `InterpExpr_Pure` (which give us our
induction hypotheses):

```dafny
lemma InterpExpr_Pure ... {
  match e
    case Seq(e1, e2) =>
      var res1 := InterpExpr(e1, ctx); // Duplicates the interpreter
      InterpExpr_Pure(e1, locals, ctx); // *Recursive call*

      if res1.Success? { // Not even in the interpreter (subsumed by the assignment with `:-`)!
        var (_, ctx1) := res1.value; // Duplicates the interpreter
        InterpExpr_Pure(e2, locals, ctx1); // *Recursive call*
      }
      else {} // Trivial case

    case ... =>
      ...
}
```

As a result, the proof above basically duplicates the code of the interpreter. This is
true for all the cases of `Expr`, and has to be done for all the lemmas we prove about the
interpreter, leading to obvious scalability issues.

> **Note:** the full proof without the induction principle is defined in the [`PureNoInductionPrinciple`](induction-principle/PureNoInductionPrinciple.dfy)

Reviewer 2: ```suggestion
**Note:** The full proof without the induction principle is defined in the [`PureNoInductionPrinciple`](induction-principle/PureNoInductionPrinciple.dfy)
```

  ensures forall st :: P(st, e)
  {
    match e
      case Seq(e1, e2) =>
        Induct(e1); // Recursion
        Induct(e2); // Recursion
        Induct_Seq(e1, e2); // Induction step for `Seq`

      // Other cases: omitted
      case ... =>
        ...
  }
```

In the case of `InterpExpr_Pure`, we would instantiate the `Induction` module in
the following manner:

```dafny
module Pure refines Induction {
  // The state contains the context, of course, but also the set of local variables
  // found in the outer bindings.
  type S = State(locals: set<string>, ctx: Context)

  // Put aside the fact that we use a state of type ``S``, this is the same predicate as ``P`` above
  predicate P ... {
    var (locals, ctx) := st;
    IsPure(e, locals) ==>
    ctx.Keys >= locals ==>
    match InterpExpr(e, ctx)
      case Success((_, ctx')) =>
        && ctx'.Keys == ctx.Keys
        && forall x | x in ctx.Keys && x !in locals :: ctx'[x] == ctx[x]
      case Failure => true
  }

  lemma Induct_Seq ... {} // Goes through automatically!

  ... // Other cases: omitted

  // And we get the theorem for all `e` for free!
}
```

### Refined induction principle - verifying `Pure`

> Stating an induction principle the way we did above allows to factorize a lot of work,

Reviewer 2: ```suggestion
Stating an induction principle the way we did above enables factorizing a lot of work,
```

manipulate the intermediate states, we need to talk about execution success and
failures. Going back to the specific proof of
[`InterpExpr_Pure`](#interpexpr-pure), we split the
[`P`](#p-predicate) predicate as follows:

```dafny
predicate P_Succ(locals: set<string>, ctx: Context, e: Expr, ctx': Context, v: int)
  // Success: the preconditions are true and we successfully evaluate `e` to
  // a pair (value, new state)
{
  && IsPure(e, locals)
  && ctx.Keys >= locals
  && InterpExpr(e, ctx) == Success((v, ctx'))
  && ctx'.Keys == ctx.Keys
  && forall x | x in ctx.Keys && x !in locals :: ctx'[x] == ctx[x]
}

predicate P_Fail(locals: set<string>, ctx: Context, e: Expr)
  // Failure: the preconditions are false or we fail to evaluate `e`
{
  IsPure(e, locals) ==>
  ctx.Keys >= locals ==>
  InterpExpr(e, ctx).Failure?
}
```

`P_Succ` and `P_Fail` must satisfy the following important property:

```dafny
lemma P_ExcludedMiddle(locals: set<string>, ctx: Context, e: Expr)
  ensures P(locals, ctx, e) <==> P_Fail(locals, ctx, e) || exists ctx', v :: P_Succ(locals, ctx, e, ctx', v)
```

<br>

> **Note:** in the generic induction principle, we actually don't request the property `P_ExcludedMiddle`

Reviewer 2: ```suggestion
**Note:** In the generic induction principle, we actually don't request the property `P_ExcludedMiddle`
```

```dafny
predicate method VarUnchanged(x: string, e: Expr)
  // Returns true if no assignments of `x` (not shadowed by a let-binding) appears
  // in `e`.
{
  match e
    case Bind(bvar, bval, body) =>
      // The rhs doesn't update x
      VarUnchanged(x, bval) &&
      // If the binding doesn't shadow x, the body of the let-binding doesn't update x
      (bvar == x || VarUnchanged(x, body))

    case Assign(avar, aval) =>
      // The variable we update is not x
      avar != x
      // And we don't update x in the rhs
      && VarUnchanged(x, aval)

    // Other cases omitted
    case ... =>
      ...
}

```

The property we want to prove is the following one:
```dafny
forall x, ctx, e ::
  VarUnchanged(x, e) ==>
  x in ctx ==>
  match InterpExpr(e, st.ctx)
    case Success((_, ctx')) =>
      && x in ctx'.Keys // x is also in the new environment
      && ctx'[x] == ctx[x] // x maps to the same value as before
    case Failure =>
      true
  Pre(st, e) ==> ResultSameX(st, res)
```

Because of the potential shadowing appearing in the `Bind` case, we actually can't use this
property directly in the inductive proof. We actually have to make the `x` string optional

> (see the [code development](induction-principle/VarUnchanged.dfy) for the full details). Once this detail properly handled, all the

Reviewer 2: ```suggestion
(see the [code development](induction-principle/VarUnchanged.dfy) for the full details). Once this detail is properly handled, all the
```

proofs go through automatically.

<a id="section-increments"></a>

## Incrementally expanding the language

We now study how incremental updates to the AST impact the induction
principle and the proofs we have done so far.

<a id="section-increments-seq"></a>

### A more general version of `Seq`: `Seq(expr, expr) ---> Seq(seq<Expr>)`

The first modification we do is to transform the `Seq` case from `Seq(e1: Expr, e2: Expr)`
to `Seq(seq<Expr>)`. This requires adding a function to interpret sequences of expressions.

> Anticipating a bit on the next language extension we will implement (allowing assignments
> and let-bindings to update/declare a sequence of variables -
> see [next section](#section-increments-varseq)), we define this function so that it returns

Reviewer 2: ```suggestion
Anticipating a bit on the next language extension we will implement allowing assignments
and let-bindings to update/declare a sequence of variables -
see [next section](#section-increments-varseq); we define this function so that it returns
```

the *sequence* of the values the individual expressions evaluate to, rather than the value
the *last* expression in the sequence evaluates to:

```dafny
function method InterpExprs(es: seq<Expr>, ctx: Context): (r:Result<(seq<int>, Context)>)
  ensures r.Success? ==> |r.value.0| == |es|
  decreases es, 0
{
  if es == [] then Success(([], ctx))
  else
    var (v, ctx1) :- InterpExpr(es[0], ctx);      // Evaluate the first expression of the sequence
    var (vs, ctx2) :- InterpExprs(es[1..], ctx1); // Evaluate the rest of the sequence
    Success(([v] + vs, ctx2))
}
```

### Updating the induction principle
Because of this new `InterpExprs` function, we need to add new predicates to the induction principles.
We previously introduced `P`, `P_Succ` and `P_Fail` to state properties about `InterpExpr`,
we now need to introduce `Pes`, `Pes_Succ` and `Pes_Fail`, to state properties about `InterpExprs`.
Similarly to the `P` predicates which manipulate values of type `V`, the `Pes` predicates
manipulate sequences of values of type `VS`.

> **Note:** if you wonder why we parameterize the induction principle by `V` and `VS`,

Reviewer 2: ```suggestion
**Note:** If you wonder why we parameterize the induction principle by `V` and `VS`,
```

predicates (not shown here).

Finally, because `VS` is not a concrete type, we need some constants and lemmas to be provided
by the user.
```dafny
function AppendValue(v: V, vs: VS): VS // Returns: [v] + vs

ghost const NilVS: VS // Empty sequence of values

function VS_Last(vs: VS): V // Last element of a non-empty sequence of values
  requires vs != NilVS
```

The induction lemmas for the `Seq` case are straightforward to update: the only important
thing is that they now sometimes use the `Pes` predicates instead of the `P` predicates
(we don't show those lemmas here).

The last elements we need to add are induction lemmas for `InterpExprs` itself. We define
them as follows:

```dafny
lemma InductExprs_Nil(st: S)
  ensures !Pes_Fail(st, []) ==> Pes_Succ(st, [], st, NilVS)

lemma InductExprs_Cons(st: S, e: Expr, es: seq<Expr>)
  ensures P_Fail(st, e) ==> Pes_Fail(st, [e] + es)
  ensures !P_Fail(st, e) ==> forall st1, v :: P_Succ(st, e, st1, v) && Pes_Fail(st1, es) ==> Pes_Fail(st, [e] + es)
  ensures forall st1, v, st2, vs :: P_Succ(st, e, st1, v) && Pes_Succ(st1, es, st2, vs) ==> Pes_Succ(st, [e] + es, st2, AppendValue(v, vs))
```

> **Note**: we grouped several cases in `InductExprs_Cons` and used forall quantifiers. The reason

Reviewer 2: ```suggestion
**Note**: We grouped several cases in `InductExprs_Cons` and used forall quantifiers. The reason
```

is that those proofs are rarely a problem: they are very similar to the `if then else`
case, in the sense that we often only need to have the induction hypotheses properly instantiated
in the context, and the context often remains small. We do however want to split cases and control
the context in cases like `Bind`, which often require a fair amount of work for
non-trivial proofs (e.g., when studying variable inlining).

### Updating the proofs

> Interestingly, the proofs for `Pure`, and `EliminateMulZero` only require
> to introduce new definitions (for `Pes` for instance). They do not require to actually
> update broken proofs. In addition, proving the `InductExprs_Cons` lemma for `VarUnchanged`

Reviewer 2: ```suggestion
Interestingly, the proofs for `Pure`, and `EliminateMulZero` only require
introducing new definitions (for `Pes` for instance). They do not require actually
updating broken proofs. In addition, proving the `InductExprs_Cons` lemma for `VarUnchanged`
```

> Interestingly, the proofs for `Pure`, and `EliminateMulZero` only require
> to introduce new definitions (for `Pes` for instance). They do not require to actually
> update broken proofs. In addition, proving the `InductExprs_Cons` lemma for `VarUnchanged`

Reviewer 2: ```suggestion
Interestingly, the proofs for `Pure`, and `EliminateMulZero` only require
introducing new definitions (for `Pes` for instance). They do not require actually
updating broken proofs. In addition, proving the `InductExprs_Cons` lemma for `VarUnchanged`
```

also requires writing an assertion to help the SMT solver.

> The proof of correctness of `IsPure` which doesn't use the induction principle

Reviewer 2: ```suggestion
The proof of correctness of `IsPure` that doesn't use the induction principle
```

(`PureNoInductionPrinciple`) is a bit more tedious, because it requires changing the `Seq`
case and introducing a proof about `InterpExprs`.

<a id="section-increments-varseq"></a>

### Declaring or updating several variables at once: `var x, y, w := 1, 2, 3;`

Now that we introduced a way of reasoning about sequences of expressions, we can update
`Bind` and `Assign` to allow updating/introducing several variables at once! We do the following
updates in the AST:

```dafny
Bind(bvar: string, bval: Expr, body: Expr)   -->   Bind(bvars: seq<string>, bvals: seq<Expr>, body: Expr)
Assign(avar: string, aval: Expr)             -->   Assign(avars: seq<string>, avals: seq<Expr>)
```

> **Note:** in the [code development](induction-principle/AST.dfy), we actually rename `Expr` to `Expr_Raw` and define

Reviewer 2: ```suggestion
**Note:** In the [code development](induction-principle/AST.dfy), we actually rename `Expr` to `Expr_Raw` and define
```

`Expr` as a subset type of `Expr_Raw` to enforce the fact that in the `Assign` and `Bind`
cases, we have the same number of variables in the left-hand side as of expressions in
the righ-hand side.

Updating the interpreter and the induction principle is then
straightforward. Interestingly, the proofs using the induction principle (`VarUnchanged`,
`Pure` and `EliminateMulZero`) only require updates to the definitions declared (and not
defined) by the induction principle: none of the proofs break.

As expected, the `PureNoInductionPrinciple` proof requires the `Assign` and `Bind` cases
to be updated.

> **Note:** if you look at the number of lines of code, you will notice that updating or

Reviewer 2: ```suggestion
**Note:** If you look at the number of lines of code, you will notice that updating or
```

providing the definitions required by the induction principle forces us to modify more lines
of code that updating the proofs when not using the induction principle. The question then
is how straightforward it is to update or provide those definitions, compared to the cost
of updating broken proofs when we don't use the induction principle. So far, we experienced

> more pain when updating the proofs. It is also worth mentionning the fact that, in the

Reviewer 2: ```suggestion
more pain when updating the proofs. It is also worth mentioning the fact that, in the
```

and their body:

```dafny
datatype Value
| Integer(i: int)
| Closure(ctx: Context, vars: seq<string>, body: Expr)
```

The problem when verifying compiler transformations is that we can't use Dafny's native
equality anymore, because local rewritings might modify the body of closures and hence their
values: we can't reason with a *syntactic* equality. We thus need to introduce a *semantic*
equality over values. This semantic equality would state that two closures are equal if,
when called on equal inputs (in the sense of this semantic equality), they evaluate
to equal results (in the sense of this semantic equality, again), in the spirit of the
following predicate:

```dafny
predicate EqValue_Closure(cv: Closure, cv': Closure)
{
  && |cv.vars| == |cv'.vars|
  && forall inputs, inputs' |
       && |inputs| == |inputs'| == |cv.vars|
       && (forall i | 0 <= i < |inputs| :: EqValue(inputs[i], inputs'[i])) ::
       var ctx := InitializeClosureCtx(cv.vars, inputs);
       var ctx' := InitializeClosureCtx(cv'.vars, inputs');
       EqResult(InterpExpr(cv.body, ctx), InterpExpr(cv'.body, ctx'))
}
```

> **Note:** this predicate is simplified for the purpose of illustration. For the "real"

Reviewer 2: ```suggestion
**Note:** This predicate is simplified for the purpose of illustration. For the "real"
```

> **Note:** this predicate is simplified for the purpose of illustration. For the "real"

Reviewer 2: ```suggestion
**Note:** This predicate is simplified for the purpose of illustration. For the "real"
```
